geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/filters/CqlTransformFilter.scala

```diff
@@ -0,0 +1,346 @@
+    extends org.apache.hadoop.hbase.filter.Filter {
+
+  /**
+    * A filter can expect the following call sequence:
+    *
+    * <ul>
+    * <li> `reset()` : reset the filter state before filtering a new row. </li>
+    * <li> `filterAllRemaining()`: true means row scan is over; false means keep going. </li>
+    * <li> `filterRowKey(byte[],int,int)`: true means drop this row; false means include.</li>
+    * <li> `filterKeyValue(Cell)`: decides whether to include or exclude this KeyValue. See `ReturnCode`. </li>
+    * <li> `transform(KeyValue)`: if the KeyValue is included, let the filter transform the KeyValue. </li>
+    * <li> `filterRowCells(List)`: allows direct modification of the final list to be submitted
+    * <li> `filterRow()`: last chance to drop entire row based on the sequence of
+    *   filter calls. Eg: filter a row if it doesn't contain a specified column. </li>
+    * </ul>
+    */
+
+  override def filterKeyValue(v: Cell): ReturnCode = delegate.filterKeyValue(v)
+  override def transformCell(v: Cell): Cell = delegate.transformCell(v)
+
+  override def reset(): Unit = {}
+  override def filterRowKey(buffer: Array[Byte], offset: Int, length: Int): Boolean = false
+  override def filterAllRemaining(): Boolean = false
+  override def transform(currentKV: KeyValue): KeyValue = currentKV
+  override def filterRowCells(kvs: java.util.List[Cell]): Unit = {}
+  override def hasFilterRow: Boolean = false
+  override def filterRow(): Boolean = false
+  override def getNextKeyHint(currentKV: KeyValue): KeyValue = null
+  override def getNextCellHint(currentKV: Cell): Cell = null
+  override def isFamilyEssential(name: Array[Byte]): Boolean = false
+  override def toByteArray: Array[Byte] = serialized
+
+  // overrides package-private method in Filter
+  protected [filters] def areSerializedFieldsEqual(other: org.apache.hadoop.hbase.filter.Filter): Boolean = true
+
+  override def toString: String = delegate.toString
+}
+
+object CqlTransformFilter extends StrictLogging {
+
+  import org.locationtech.geomesa.index.FilterCacheSize
+
+  val Priority: Int = 30
+
+  private val cache = Caffeine.newBuilder().maximumSize(FilterCacheSize.toInt.get).build(
+    new CacheLoader[ByteArrayCacheKey, DelegateFilter]() {
+      override def load(key: ByteArrayCacheKey): DelegateFilter = {
+        val filter = deserialize(key.bytes)
+        logger.trace(s"Deserialized $filter")
+        filter
+      }
+    }
+  )
+
+  /**
+    * Override of static method from org.apache.hadoop.hbase.filter.Filter
+    *
+    * @param pbBytes serialized bytes
+    * @throws org.apache.hadoop.hbase.exceptions.DeserializationException serialization exception
+    * @return
+    */
+  @throws(classOf[DeserializationException])
+  def parseFrom(pbBytes: Array[Byte]): org.apache.hadoop.hbase.filter.Filter =
+    // note: we copy KryoBufferSimpleFeatures as they seem to cause problems if re-used, even if thread-local
```

Comment:
We are copying something other than the KryoBufferSF here, right?  

Also, if we have to copy, is the cache worth it?

Reply:
yeah, we're instantiating a new instance of the delegate filter, but it should be very lightweight. the KryoBufferSimpleFeature copy is slightly more expensive, as it will re-run some code to generate the transform function [here](https://github.com/locationtech/geomesa/blob/master/geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala#L120-L185) (if there are transforms set). But it saves creating a new serializer, and some cache lookups for getting the readers/writers... I think it's probably worthwhile, and it would make it easy to optimize further if we can figure out a way to fully re-use the features (possibly some kind of object pool, but that seemed slower when we looked at other pieces of the code that were using them).
Originally I was re-using the same KryoBufferSimpleFeature, but for some reason that was causing occasional thread conflicts, so it seems like each filter instance needs to have a separate copy